Add design diagram and notes on code, for easier development (#158)
kubernetes-sigs/cluster-api-provider-azure: docs/design.md

# Cluster API Provider Azure Design<!-- omit in toc -->

![Design](./design.jpg)

> ## Folder Structure

Reviewer: Let's remove this; we'll have to keep track of it every time something changes.

Author: Yes, we should keep track of the design changes, and any future reviewers would be able to just refer this instead of unaware of why it was done a certain way. I prefer to keep this for now

* /pkg/cloud/azure/apis - contains the crds
* /pkg/cloud/azure
  * /pkg/cloud/azure/actuators/cluster - cluster actuator for cluster api provider
  * /pkg/cloud/azure/actuators/machine - machine actuator for cluster api provider
  * /pkg/cloud/azure/services - each folder represents an azure service, implements Get/CreateOrUpdate/Delete
  * /pkg/cloud/azure/services/config - contains startup scripts for controlplane and worker nodes

## Components

### Cluster Actuator

> Cluster actuator creates the following in order

Reviewer: ```suggestion
The cluster actuator creates the following resources, in order:
```

> 

Reviewer: Remove line

> * Certificates using kubeadm generating 

Reviewer: Prefer hyphens over asterisks for lists.

> * Certificates using kubeadm generating 

Reviewer: ```suggestion
- Certificates (generating using the kubeadm packages): 
```

>     * CA

Reviewer: Let's prefer expository over brevity when describing the resources.

>     * CA

Reviewer: ```suggestion
    - Kubernetes Certificate Authority (CA)
```

>     * Api-server uses DNS Name as input

Reviewer: ```suggestion
    - API server (includes public IP FQDN as a certificate SAN (subject alternative name))
```

>     * Front-Proxy

Reviewer: ```suggestion
    - front proxy
```

>     * SA

Reviewer: ```suggestion
    - SA (Service account)
```

>     * Admin Kubeconfig

Reviewer: ```suggestion
    - Admin kubeconfig
```

>     * Discovery Hashes required for bootstrapping

Reviewer: ```suggestion
    - kubeadm discovery hashes (required for bootstrapping)
```

> * Zone redundant Base Infrastructure

Reviewer: ```suggestion
- Azure networking components (zone-redundant, by default)
```

>     * Vnet

Reviewer: ```suggestion
    - (1) virtual network
```

>     * Subnets

Reviewer: ```suggestion
    - (2) subnets
```

>       * ControlPlane

Reviewer: ```suggestion
      - control plane
```

>       * Node

Reviewer: ```suggestion
      - node
```

>     * NSG

Reviewer: ```suggestion
    - (2) NSG (network security groups)
```

>       * ControlPlane

Reviewer: ```suggestion
      - control plane
```

>       * Node

Reviewer: ```suggestion
      - node
```

>     * Node Routetable

Reviewer: ```suggestion
    - (1) route table (node)
```

>     * Standard Public IP

Reviewer: ```suggestion
    - (1) public IP (Standard SKU) 
```

>     * Standard Public loadbalancer using public ip

Reviewer 2: Super nit here, can you make `ip` `IP`

>     * Standard Public loadbalancer using public ip

Reviewer: ```suggestion
    - (1) public load balancer (Standard SKU, attached to public IP)
```

>     * Standard Internal loadbalancer for communication from worker nodes

Reviewer: ```suggestion
   - (1) internal load balancer (Standard SKU, for communication from worker nodes)
```

### Machine Actuator

> Machine actuator creates control plane or worker nodes using above base infrastructure, in following order

Reviewer: ```suggestion
The machine actuator creates Kubernetes nodes, according to the supplied `machines.yaml`, within the above Azure network infrastructure, in the following order:
```

> 

Reviewer: Delete line.

> * Network Interface

Reviewer: ```suggestion
- NIC (network interface card)
```

> * Virtual Machine in an Availability Zone

Reviewer: ```suggestion
- virtual machine (created in an Azure Availability Zone, in regions where AZs are supported)
```

> * Generates Bootstrap token if required and generates startup scripts

Reviewer: Separate lines for the bootstrap token and startup script

> * Generates Bootstrap token if required and generates startup scripts

Reviewer: ```suggestion
- kubeadm bootstrap token (if required)
```

> * Custom script extension using above startup script

Reviewer: ```suggestion
- generated startup script (to be used by Azure Custom Script Extension)
```
